src/diem/offchain/__init__.py

```diff
@@ -37,14 +44,20 @@
     to_json,
     from_json,
     validate_write_once_fields,
+    FundPullPreApprovalType,
+    new_funds_pull_pre_approval_object,
 )
 from .http_header import X_REQUEST_ID, X_REQUEST_SENDER_ADDRESS
 from .action import Action
 from .error import command_error, protocol_error, Error
 from .command import Command
 from .payment_command import PaymentCommand
+from .funds_pull_pre_approval_command import FundsPullPreApprovalCommand
+
+import typing
+
+CommandVariant = typing.Union[FundsPullPreApprovalCommand, PaymentCommand]
```

Comment:
> FundsPullPreApprovalCommand and PaymentCommand are inherit from Command, should not need another Union type.

The rationale here is that this type is used in `process_inbound_request` – a method that doesn't rely on the polymorphic properties of the command objects; it actually has separate code section for each command type. As such, marking it as returning Command both adds unnecessary constraints to the method and conceals information that tells a more correct story to the user – this method doesn't return just some arbitrary subclass of Command, leaving the user to guess which variants are viable, it actually returns one of the two options, disregarding their inheritance. Union is a semantically richer option that doesn't bring with it the need for specific ancestry where it doesn't add descriptive value.

Bottom line, being type hints, it is not super critical in my opinion, the code will work the same in both cases. I do believe a union is the better option, but ultimately it is your call @xli.

Reply:
It depends on how you use it. instead of use CommandVariant union everywhere, use Command with interface is better for the case of caller only need to know the Command interface.
Eventually, no where should check what is sub-class of the Command.

Reply:
> Eventually, no where should check what is sub-class of the Command.

I don't think we have these checks anywhere.

>  Command with interface is better for the case of caller only need to know the Command interface.

The caller is not forced to know anything and is free to use the base class or the concrete class as he wishes. The type notation doesn't affect the calling code or the code of `process_inbound_request`. It does give a more precise information though.

Anyhow, how would you prefer us to proceed with it? Would you prefer we change the places using the union to use the base class?

Reply:
Use base class if possible, union type gives an impression that the types united don't have same interface.